File: backend/src/services/entryService.ts
import type { MealType } from "@prisma/client";
import { exclusiveEnd, startOfUtcDay } from "../lib/dates.js";
import { NotFoundError, ValidationError } from "../lib/errors.js";
import { mapEntry } from "../lib/mappers.js";
import { paginated, parsePagination } from "../lib/pagination.js";
import { entryRepo } from "../repositories/entryRepo.js";
import type { CreateEntryInput, UpdateEntryInput } from "../validators/entries.js";

function parseConsumedAt(value: string): Date {
  try {
    return /^\d{4}-\d{2}-\d{2}$/.test(value) ? startOfUtcDay(value) : new Date(value);
  } catch {
    throw new ValidationError("consumedAt must be an ISO date or datetime");
  }
}

function parseRange(from: string, to: string): { from: Date; toExclusive: Date } {
  let start: Date;
  let end: Date;
  try {
    start = startOfUtcDay(from);
    end = exclusiveEnd(to);
  } catch {
    throw new ValidationError("from and to must be ISO dates or datetimes");
  }
  if (end <= start) {
    throw new ValidationError("`to` must be after `from`");
  }
  return { from: start, toExclusive: end };
}

export const entryService = {
  async list(
    userId: string,
    query: {
      from: string;
      to: string;
      mealType?: MealType;
      limit?: unknown;
      offset?: unknown;
    },
  ) {
    const pagination = parsePagination(query);
    const range = parseRange(query.from, query.to);
    const [rows, total] = await entryRepo.list(
      {
        userId,
        from: range.from,
        toExclusive: range.toExclusive,
        mealType: query.mealType,
      },
      pagination.limit,
      pagination.offset,
    );
    return paginated(rows.map(mapEntry), total, pagination);
  },

  async getById(userId: string, id: string) {
    const entry = await entryRepo.findById(id);
    if (!entry || entry.userId !== userId) {
      throw new NotFoundError("Food entry not found");
    }
    return mapEntry(entry);
  },

  async create(userId: string, input: CreateEntryInput) {
    const entry = await entryRepo.create({
      userId,
      mealType: input.mealType,
      foodName: input.foodName,
      quantity: input.quantity,
      servingUnit: input.servingUnit,
      calories: input.calories,
      proteinGrams: input.proteinGrams,
      carbsGrams: input.carbsGrams,
      fatGrams: input.fatGrams,
      vitaminCMg: input.vitaminCMg ?? 0,
      calciumMg: input.calciumMg ?? 0,
      ironMg: input.ironMg ?? 0,
      vitaminDIU: input.vitaminDIU ?? 0,
      potassiumMg: input.potassiumMg ?? 0,
      consumedAt: parseConsumedAt(input.consumedAt),
    });
    return mapEntry(entry);
  },

  async update(userId: string, id: string, input: UpdateEntryInput) {
    const existing = await entryRepo.findById(id);
    if (!existing || existing.userId !== userId) {
      throw new NotFoundError("Food entry not found");
    }

    const entry = await entryRepo.update(id, {
      ...(input.mealType !== undefined ? { mealType: input.mealType } : {}),
      ...(input.foodName !== undefined ? { foodName: input.foodName } : {}),
      ...(input.quantity !== undefined ? { quantity: input.quantity } : {}),
      ...(input.servingUnit !== undefined ? { servingUnit: input.servingUnit } : {}),
      ...(input.calories !== undefined ? { calories: input.calories } : {}),
      ...(input.proteinGrams !== undefined ? { proteinGrams: input.proteinGrams } : {}),
      ...(input.carbsGrams !== undefined ? { carbsGrams: input.carbsGrams } : {}),
      ...(input.fatGrams !== undefined ? { fatGrams: input.fatGrams } : {}),
      ...(input.vitaminCMg !== undefined ? { vitaminCMg: input.vitaminCMg } : {}),
      ...(input.calciumMg !== undefined ? { calciumMg: input.calciumMg } : {}),
      ...(input.ironMg !== undefined ? { ironMg: input.ironMg } : {}),
      ...(input.vitaminDIU !== undefined ? { vitaminDIU: input.vitaminDIU } : {}),
      ...(input.potassiumMg !== undefined ? { potassiumMg: input.potassiumMg } : {}),
      ...(input.consumedAt !== undefined ? { consumedAt: parseConsumedAt(input.consumedAt) } : {}),
    });
    return mapEntry(entry);
  },

  async remove(userId: string, id: string) {
    const existing = await entryRepo.findById(id);
    if (!existing || existing.userId !== userId) {
      throw new NotFoundError("Food entry not found");
    }
    await entryRepo.delete(id);
  },
};

File: backend/src/main/java/com/nourish/tracker/service/EntryService.java
package com.nourish.tracker.service;

import com.nourish.tracker.api.common.PageResponse;
import com.nourish.tracker.api.entry.CreateEntryRequest;
import com.nourish.tracker.api.entry.EntryResponse;
import com.nourish.tracker.api.entry.UpdateEntryRequest;
import com.nourish.tracker.api.error.ApiException;
import com.nourish.tracker.domain.FoodEntry;
import com.nourish.tracker.domain.MealType;
import com.nourish.tracker.repository.FoodEntryRepository;
import com.nourish.tracker.support.OffsetLimitPageable;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.UUID;

@Service
public class EntryService {
    private final FoodEntryRepository entryRepository;
    private final DemoUserService demoUserService;

    public EntryService(
            FoodEntryRepository entryRepository,
            DemoUserService demoUserService
    ) {
        this.entryRepository = entryRepository;
        this.demoUserService = demoUserService;
    }

    @Transactional
    public EntryResponse create(CreateEntryRequest request) {
        FoodEntry entry = new FoodEntry();
        entry.setUser(demoUserService.getDemoUser());
        applyCreateRequest(entry, request);
        return EntryResponse.from(entryRepository.save(entry));
    }

    @Transactional(readOnly = true)
    public PageResponse<EntryResponse> list(
            LocalDate from,
            LocalDate to,
            MealType mealType,
            int limit,
            long offset
    ) {
        if (to.isBefore(from)) {
            throw new ApiException(
                    HttpStatus.BAD_REQUEST,
                    "INVALID_DATE_RANGE",
                    "to must be on or after from"
            );
        }

        Instant fromInstant = from.atStartOfDay().toInstant(ZoneOffset.UTC);
        Instant toExclusive = to.plusDays(1).atStartOfDay().toInstant(ZoneOffset.UTC);
        var pageable = new OffsetLimitPageable(offset, limit);

        Page<FoodEntry> page = mealType == null
                ? entryRepository
                .findByUserIdAndConsumedAtGreaterThanEqualAndConsumedAtLessThanOrderByConsumedAtDesc(
                        DemoUserService.DEMO_USER_ID,
                        fromInstant,
                        toExclusive,
                        pageable
                )
                : entryRepository
                .findByUserIdAndMealTypeAndConsumedAtGreaterThanEqualAndConsumedAtLessThanOrderByConsumedAtDesc(
                        DemoUserService.DEMO_USER_ID,
                        mealType,
                        fromInstant,
                        toExclusive,
                        pageable
                );

        return PageResponse.from(page, limit, offset, EntryResponse::from);
    }

    @Transactional(readOnly = true)
    public EntryResponse get(UUID id) {
        return EntryResponse.from(findOwned(id));
    }

    @Transactional
    public EntryResponse update(UUID id, UpdateEntryRequest request) {
        FoodEntry entry = findOwned(id);

        if (request.mealType() != null) entry.setMealType(request.mealType());
        if (request.foodName() != null) entry.setFoodName(request.foodName().trim());
        if (request.quantity() != null) entry.setQuantity(request.quantity());
        if (request.servingUnit() != null) entry.setServingUnit(request.servingUnit().trim());
        if (request.calories() != null) entry.setCalories(request.calories());
        if (request.proteinGrams() != null) entry.setProteinGrams(request.proteinGrams());
        if (request.carbsGrams() != null) entry.setCarbsGrams(request.carbsGrams());
        if (request.fatGrams() != null) entry.setFatGrams(request.fatGrams());
        if (request.vitaminCMg() != null) entry.setVitaminCMg(request.vitaminCMg());
        if (request.calciumMg() != null) entry.setCalciumMg(request.calciumMg());
        if (request.ironMg() != null) entry.setIronMg(request.ironMg());
        if (request.vitaminDIU() != null) entry.setVitaminDIU(request.vitaminDIU());
        if (request.potassiumMg() != null) entry.setPotassiumMg(request.potassiumMg());
        if (request.consumedAt() != null) entry.setConsumedAt(request.consumedAt());

        return EntryResponse.from(entryRepository.save(entry));
    }

    @Transactional
    public void delete(UUID id) {
        entryRepository.delete(findOwned(id));
    }

    private void applyCreateRequest(FoodEntry entry, CreateEntryRequest request) {
        entry.setMealType(request.mealType());
        entry.setFoodName(request.foodName().trim());
        entry.setQuantity(request.quantity());
        entry.setServingUnit(request.servingUnit().trim());
        entry.setCalories(request.calories());
        entry.setProteinGrams(request.proteinGrams());
        entry.setCarbsGrams(request.carbsGrams());
        entry.setFatGrams(request.fatGrams());
        entry.setVitaminCMg(orZero(request.vitaminCMg()));
        entry.setCalciumMg(orZero(request.calciumMg()));
        entry.setIronMg(orZero(request.ironMg()));
        entry.setVitaminDIU(orZero(request.vitaminDIU()));
        entry.setPotassiumMg(orZero(request.potassiumMg()));
        entry.setConsumedAt(request.consumedAt());
    }

    private FoodEntry findOwned(UUID id) {
        return entryRepository.findById(id)
                .filter(entry -> DemoUserService.DEMO_USER_ID.equals(entry.getUser().getId()))
                .orElseThrow(() -> new ApiException(
                        HttpStatus.NOT_FOUND,
                        "ENTRY_NOT_FOUND",
                        "Food entry was not found"
                ));
    }

    private BigDecimal orZero(BigDecimal value) {
        return value == null ? BigDecimal.ZERO : value;
    }
}
